pkg/client/client.go

```diff
@@ -0,0 +1,72 @@
+package client
+
+import (
+	"context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+// ObjectKey identifies a Kubernetes Object.
+type ObjectKey types.NamespacedName
+
+// Client uses methods defined in runtime.Object to determine the REST endpoint
+// for Kubernetes resources.
+
+// Interface defines the capability for a Kubernetes client.
+type Interface interface {
+	// Create saves the object obj in the Kubernetes cluster.
+	Create(ctx context.Context, obj runtime.Object) error
+
+	// Get retrieves an obj for the given object key from the Kubernetes Cluster.
+	// obj must be a struct pointer so that obj can be updated with the response
+	// returned by the Server.
+	Get(ctx context.Context, key ObjectKey, obj runtime.Object) error
+
+	// Update updates the given obj in the Kubernetes cluster. obj must be a
+	// struct pointer so that obj can be updated with the content returned by the Server.
+	Update(ctx context.Context, obj runtime.Object) error
+
+	// Delete deletes the given obj from Kubernetes cluster.
+	Delete(ctx context.Context, obj runtime.Object) error
+
+	// List retrieves list of objects for a given namespace and list options. On a
+	// successful call, Items field in the list will be populated with the
+	// result returned from the server.
+	List(ctx context.Context, namespace string, opts *metav1.ListOptions, list runtime.Object) error
```

Comment:
What is the reason why you're modifying the list in-place? instead of returning it?

Reply:
Same reasoning as above. The hidden element here is that `obj` gives us `gvk` input which is causing this issue.

Reply:
Can you create a GVKN and have a method to parse an object into a GVKN?
At this point, you barely have two methods that are consistent in usage, which is very confusing. I've always been confused in the past by libraries that took empty object to guess what to do.

Reply:
One of the aims was to not have to specify a GVK when we didn't need to/it was redundant, and to not require type assertions.  The user already knows what kind of object they're using, it seems silly to have them write that twice.

The actual output code is fairly readable, IMO:

```go
var pods kapi.PodList
if err := client.List(context.TODO(), "somenamespace", nil, &pods); err != nil {
   return err
}
```

as opposed to with GVK:

```
obj, err := client.List(context.TODO(), "foo", nil, schema.GroupVersionKind{Version: "v1", Kind: "Pod")
if err != nil {
   return err
}
podList := obj.(*kapi.PodList) // hope that this is correct and that you didn't make a typo
```